packages/nimble-components/src/table/specs/table-column-specs/table-column-mapping.md

```diff
@@ -0,0 +1,324 @@
+# Mapping Table Column
+
+## Overview
+
+The `nimble-table-column-mapping` is a component that supports rendering specific number, boolean, or string values as mapped text. `nimble-table-column-icon` is a specialized version of `nimble-table-column-mapping` that instead maps values to icons and/or spinners and has a minimal, fixed width. The actual mappings are defined by child elements `nimble-mapping-icon`, `nimble-mapping-spinner`, and `nimble-mapping-text`.
+
+### Background
+
+[Icon column type issue](https://github.com/ni/nimble/issues/1013)
+
+[Boolean column type issue](https://github.com/ni/nimble/issues/1103)
+
+### Features
+
+-   Supported input:
+    -   string
+    -   number
+    -   boolean
+-   Supported output:
+    -   Text
+    -   Icon
+    -   Spinner
+    -   (empty)
+
+### Non-goals
+
+-   Non-Nimble icon support
+-   Arbitrary icon colors
+-   Hyperlink icons
+-   Mixed text and icons
+-   Non-icon, non-spinner Nimble components
+
+---
+
+## Design
+
+Below is an example of how these elements would be used within a `nimble-table`:
+
+```HTML
+<nimble-table>
+    <nimble-table-column-icon field-name="status">
+        Status
+        <nimble-mapping-icon key="fail" icon="nimble-icon-xmark" severity="error" label="Failed"></nimble-mapping-icon>
+        <nimble-mapping-icon key="error" icon="nimble-icon-xmark" severity="error" label="Errored"></nimble-mapping-icon>
+        <nimble-mapping-icon key="pass" icon="nimble-icon-check" severity="success" label="Passed"></nimble-mapping-icon>
+        <nimble-mapping-spinner key="running" label="Running"></nimble-mapping-spinner>
+    </nimble-table-column-icon>
+    <nimble-table-column-mapping field-name="errorCode">
+        Error Code
+        <nimble-mapping-text key="1" label="A bad thing happened"></nimble-mapping-text>
+        <nimble-mapping-text key="2" label="A worse thing happened"></nimble-mapping-text>
+        <nimble-mapping-text key="3" label="A terrible thing happened"></nimble-mapping-text>
+    </nimble-table-column-mapping>
+    <nimble-table-column-icon field-name="archived">
+        Archived
+        <nimble-mapping-icon key="true" icon="nimble-icon-database" label="Archived"></nimble-mapping-icon>
+    </nimble-table-column-icon>
+</nimble-table>
+```
+
+Note that the `key` attribute values are always given as strings. In the case of boolean or number mappings, this value is parsed to the appropriate type, so that it can properly be compared to the values from the row records.
```

Comment:
Can you elaborate on the recommendation that `key` attribute values are always strings?
1. what does this look like in client code? If I have an integer enum that I want to map to icons, do I have to explicitly call `toString()` in the template?
2. what alternatives did we consider and why did we choose this approach? Some that come to mind:
    1. mapping elements expose multiple attributes `key-string`, `key-numeric`, `key-boolean`. Nimble could enforce that only one attribute is used within a mapping column.
    2. `key` attribute accepts values of type `string | number | boolean`. Nimble could enforce that all values are the same type within a mapping column.
    3. different mapping elements for different input types. `nimble-mapping-text-string`, `nimble-mapping-text-numeric`, `nimble-mapping-text-boolean`, `nimble-mapping-icon-string`, ...

Reply:
I think maybe the point I was trying to make is too obvious, or I am just confused. I simply meant that in html, attribute values are always specified in quotes (i.e. as a string): `<element attr="value" ...`. The `key` property will be typed as `string | number | boolean`. I guess I'm not clear what type you get for `key="true"`--does typescript treat that value as the string "true" or the boolean `true`? I was assuming we would have to run type conversions from string to bool/number.

Reply:
Gotcha, I think I was also confused so thanks for the clarification.

Reply:
I think I'm confused now. The attribute will always be specified as a string but the property will be `string | number | boolean`? When will the conversion happen? Will it happen when you check the value or will it happen as part of a custom attribute to property converter that we write?

Reply:
With that direction I'd expect the column to do the conversion, the column is the one that know the target type from the field, not the mapping itself.

I think in this case we should also configure the mode of the key attribute to be `mode: 'fromView'`: https://www.fast.design/docs/fast-element/defining-elements#customizing-attributes

Since it would be ambiguous when assigned back to the attribute what the actual type of the serialized string is.

Reply:
Actually.. is that kind of pattern possible to represent in a Blazor component?

Would three nullable keys be less fiddly? `numeric-key`, `string-key`, `boolean-key`? It's spiritually similar to the [`input` element valueAsNumber, valueAsDate, etc](https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement).

Reply:
I confirmed that typing `key` as `string | number | boolean` doesn't really do anything for me. When I do `typeof mapping.key`, I always get `string`. So I think I might as well just type it as such.

My current implementation doesn't convert the value until the last moment, when it is trying to compare against the `cellView.cellRecord.value`. It's at that point that I can get the `typeof` the record value and convert the mapping's `key` value accordingly:
```ts
public isMappingFor(value: TableFieldValue): boolean {
  const dataType = typeof value;
  if (dataType === 'number') {
      return value === nullableNumberConverter.fromView(this.key);
  }
  if (dataType === 'boolean') {
      return value === booleanConverter.fromView(this.key);
  }
  if (dataType === 'string') {
      return value === this.key;
  }
  return false; // value is null or undefined
}
```

Reply:
> I confirmed that typing key as string | number | boolean doesn't really do anything for me. When I do typeof mapping.key, I always get string

TypeScript types don't exist at runtime. How the property behaves depends on if you do property assignment or attribute assignment. FAST doesn't reject property assignments that are invalid, so assigning `el.key = 5` will assign a `typeof this.key === 'number'` value. The typing is for a user using TypeScript to limit what can be assigned. For attributes it depends on the converter used what ends up in the proeprty from an attribute assignment.

If we make it only string then they have to do something like `el.key = String(5)` for TypeScript to not complain.

Reply:
> Actually.. is that kind of pattern possible to represent in a Blazor component?
> 
> Would three nullable keys be less fiddly? numeric-key, string-key, boolean-key?

I think you're right. For the sake of Blazor, I think we have to have three separately typed attributes. I considered whether we should also have a multi-typed `key` attribute to simplify the API for vanilla html and Angular, but I think it would probably just confiuse things more.

Reply:
> Actually.. is that kind of pattern possible to represent in a Blazor component?
> 
> Would three nullable keys be less fiddly? `numeric-key`, `string-key`, `boolean-key`? It's spiritually similar to the [`input` element valueAsNumber, valueAsDate, etc](https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement).

In Blazor, we would want this to be generic (ideally). So on the Blazor `NimbleMappingText` component, if I wanted my key to be an integer (because that's the type of the data), I would have:

```html
<NimbleTableColumnMapping>
    <NimbleMappingText TValue=int Key="1"></NimbleMappingText>
    <NimbleMappingText TValue=int Key="2"></NimbleMappingText>
</NimbleTableColumnMapping>
```

It would be great if `NimbleTableColumnMapping` could be the generically typed object that would then give the `NimbleMappingText` objects within an implicit typing, but we can't do that. This means that the user could also give `NimbleMappingText` objects different typing, which would have to be flagged as invalid by the table. I think we could implement the component side in a way like Milan suggests, where the component will set either `numeric-key`, `string-key`, or `boolean-key` based on the type of `TValue`. The component would throw if an invalid type was given to `TValue`.

Reply:
Does `NimbleMapping*` have to be generic if we have distinct properties `key-number`, `key-string`, and `key-boolean`?

Reply:
For Blazor we wouldn't want those distinct properties on the Blazor API surface (that is what generics are for). We can hide the `key-number`, `key-string`, and `key-boolean` attributes as implementation details (set in the component .razor code conditionally).

I'll admit that this is an example of consciously diverging the web component and Blazor APIs, but I feel that this is a good example of us changing to adopt an expected API for the platform (similar to having a property for the data instead of a method).

Reply:
That brings me back to the question of whether we actually need three distinctly typed key properties. If the Blazor component is going to have a single, generically typed key property, can we have a single key property (of type `number | string | boolean | null`) on the underlying component?

Reply:
But isn't the issue with converting the key value to the appropriate type? If you have the three distinct properties there is no confusion. How does a particular mapping column know that its `key` should be a `number` versus a `string`?

Reply:
We've come around to saying we're going to put a `data-type` attribute on the column, which means we don't really need the `key` property to specify its own type. Going back to a single `key` attr/prop.